source/Octopus.Tentacle/Configuration/Instances/ApplicationInstanceSelector.cs

```diff
@@ -64,30 +69,48 @@ ApplicationInstanceConfiguration LoadCurrentInstance()
         ApplicationInstanceConfiguration LoadInstance()
         {
             var appInstance = LocateApplicationPrimaryConfiguration();
-            EnsureConfigurationExists(appInstance.instanceName, appInstance.configurationpath);
 
-            log.Verbose($"Loading configuration from {appInstance.configurationpath}");
-            var writableConfig = new XmlFileKeyValueStore(fileSystem, appInstance.configurationpath);
-            
-            var aggregatedKeyValueStore = ContributeAdditionalConfiguration(writableConfig);
+            var (aggregatedKeyValueStore, writableConfig) = LoadConfigurationStore(appInstance);
+
             return new ApplicationInstanceConfiguration(appInstance.instanceName, appInstance.configurationpath, aggregatedKeyValueStore, writableConfig);
         }
 
-        void EnsureConfigurationExists(string? instanceName, string configurationPath)
+        (IKeyValueStore, IWritableKeyValueStore) LoadConfigurationStore((string? instanceName, string? configurationpath) appInstance)
         {
-            if (!fileSystem.FileExists(configurationPath))
+            if (appInstance is { instanceName: not null, configurationpath: null } &&
+                PlatformDetection.Kubernetes.IsRunningInKubernetes)
             {
-                var message = !string.IsNullOrEmpty(instanceName)
-                    ? $"The configuration file for instance {instanceName} could not be located at the specified location {configurationPath}. " +
-                    "The file might have been manually removed without properly removing the instance and as such it is still listed as present." +
-                    "The instance must be created again before you can interact with it."
-                    : $"The configuration file at {configurationPath} could not be located at the specified location.";
-                    
-                throw new ControlledFailureException(message);
+                log.Verbose($"Loading configuration from ConfigMap for namespace {KubernetesConfig.Namespace}");
+                var configMapWritableStore = configMapStoreFactory.Value;
+                return (ContributeAdditionalConfiguration(configMapWritableStore), configMapWritableStore);
             }
+            if (!ConfigurationFileExists(appInstance.configurationpath))
+            {
+                ThrowConfigurationMissingException(appInstance.instanceName, appInstance.configurationpath);
+            }
+
+            log.Verbose($"Loading configuration from {appInstance.configurationpath}");
+            var writable = new XmlFileKeyValueStore(fileSystem, appInstance.configurationpath!);
+            return (ContributeAdditionalConfiguration(writable), writable);
         }
 
-        AggregatedKeyValueStore ContributeAdditionalConfiguration(XmlFileKeyValueStore writableConfig)
+        bool ConfigurationFileExists([NotNullWhen(true)] string? configurationPath)
```

Comment:
nit: Did these methods need to be pulled out? It may be best to leave what we can alone to avoid any chance of introducing inadvertent changes.

ultra nit: Also since this just throws we could have pulled out the entire `if` into a single method.